src/Adapters/Storage/Mongo/MongoTransform.js

```diff
@@ -198,18 +202,114 @@ function transformWhere(schema, className, restWhere, options = {validate: true}
+    return {key: 'expiresAt', value: coercedToDate};
+  case '_id': //TODO: for some reason, _PushStatus is already transformed when it gets here. For now,
+  // just pass the ID through. Later, we should make sure the push status doesn't get transformed inside Parse Server.
+  case '_rperm':
+  case '_wperm':
+  case '_email_verify_token':
+  case '_hashed_password':
+  case '_perishable_token': return {key: restKey, value: restValue};
+  case 'sessionToken': return {key: '_session_token', value: restValue};
+  default:
+    // Auth data should have been transformed already
+    if (restKey.match(/^authData\.([a-zA-Z0-9_]+)\.id$/)) {
+      throw new Parse.Error(Parse.Error.INVALID_KEY_NAME, 'can only query on ' + restKey);
+    }
+    // Trust that the auth data has been transformed and save it directly
+    if (restKey.match(/^_auth_data_[a-zA-Z0-9_]+$/)) {
+      return {key: restKey, value: restValue};
+    }
+  }
+  //skip straight to transformAtom for Bytes, they don't show up in the schema for some reason
+  if (restValue && restValue.__type !== 'Bytes') {
+    //Note: We may not know the type of a field here, as the user could be saving (null) to a field
+    //That never existed before, meaning we can't infer the type.
+    if (parseFormatSchema.fields[restKey] && parseFormatSchema.fields[restKey].type == 'Pointer' || restValue.__type == 'Pointer') {
+      restKey = '_p_' + restKey;
+    }
+  }
+
+  // Handle atomic values
+  var value = transformAtom(restValue, false, { inArray: false, inObject: false });
+  if (value !== CannotTransform) {
+    return {key: restKey, value: value};
+  }
+
+  // ACLs are handled before this method is called
+  // If an ACL key still exists here, something is wrong.
+  if (restKey === 'ACL') {
+    throw 'There was a problem transforming an ACL.';
+  }
+
+  // Handle arrays
+  if (restValue instanceof Array) {
+    value = restValue.map((restObj) => {
+      var out = transformKeyValue(schema, className, restKey, restObj, { inArray: true });
+      return out.value;
+    });
+    return {key: restKey, value: value};
+  }
+
+  // Handle update operators
+  value = transformUpdateOperator(restValue, true);
+  if (value !== CannotTransform) {
+    return {key: restKey, value: value};
+  }
+
+  // Handle normal objects by recursing
+  value = {};
+  for (var subRestKey in restValue) {
+    var subRestValue = restValue[subRestKey];
+    var out = transformKeyValue(schema, className, subRestKey, subRestValue, { inObject: true });
+    // For recursed objects, keep the keys in rest format
+    value[subRestKey] = out.value;
+  }
+  return {key: restKey, value: value};
+}
+
 // Main exposed method to create new objects.
 // restCreate is the "create" clause in REST API form.
-// Returns the mongo form of the object.
-function parseObjectToMongoObject(schema, className, restCreate) {
+function parseObjectToMongoObjectForCreate(schema, className, restCreate, parseFormatSchema) {
```

Comment:
Just for the sake of maintenance, maybe having the 2 code paths can introduce problems. Not sure about that. I feel that having separate transforms depending on the state of the object (create, update) may introduce other issues.